sklearn/cluster/_kmeans.py

```diff
@@ -1665,160 +1662,120 @@ def fit(self, X, y=None, sample_weight=None):
+
+            # Compute inertia on a validation set.
+            _, inertia = _labels_inertia_threadpool_limit(
+                X_valid, sample_weight_valid, x_squared_norms_valid,
+                cluster_centers, n_threads=self._n_threads)
+
             if self.verbose:
-                print("Inertia for init %d/%d: %f"
-                      % (init_idx + 1, self._n_init, inertia))
+                print(f"Inertia for init {init_idx + 1}/{self._n_init}: "
+                      f"{inertia}")
             if best_inertia is None or inertia < best_inertia:
-                self.cluster_centers_ = cluster_centers
-                self._counts = weight_sums
+                init_centers = cluster_centers
                 best_inertia = inertia
 
-        # Empty context to be used inplace by the convergence check routine
-        convergence_context = {}
-
-        # Perform the iterative optimization until the final convergence
-        # criterion
-        for iteration_idx in range(n_iter):
-            # Sample a minibatch from the full dataset
-            minibatch_indices = random_state.randint(
-                0, n_samples, self.batch_size)
-
-            # Perform the actual update step on the minibatch data
-            batch_inertia, centers_squared_diff = _mini_batch_step(
-                X[minibatch_indices], sample_weight[minibatch_indices],
-                x_squared_norms[minibatch_indices],
-                self.cluster_centers_, self._counts,
-                old_center_buffer, tol > 0.0, distances=distances,
-                # Here we randomly choose whether to perform
-                # random reassignment: the choice is done as a function
-                # of the iteration index, and the minimum number of
-                # counts, in order to force this reassignment to happen
-                # every once in a while
-                random_reassign=((iteration_idx + 1)
-                                 % (10 + int(self._counts.min())) == 0),
-                random_state=random_state,
-                reassignment_ratio=self.reassignment_ratio,
-                verbose=self.verbose)
-
-            # Monitor convergence and do early stopping if necessary
-            if _mini_batch_convergence(
-                    self, iteration_idx, n_iter, tol, n_samples,
-                    centers_squared_diff, batch_inertia, convergence_context,
-                    verbose=self.verbose):
-                break
+        centers = init_centers
+        centers_new = np.empty_like(centers)
+
+        # Initialize counts
+        self._counts = np.zeros(self.n_clusters, dtype=X.dtype)
+
+        # Attributes to monitor the convergence
+        self._ewa_inertia = None
+        self._ewa_inertia_min = None
+        self._no_improvement = 0
+
+        # Initialize number of samples seen since last reassignment
+        self._n_since_last_reassign = 0
+
+        n_steps = (self.max_iter * n_samples) // self._batch_size
```

Comment:
```suggestion
        n_steps = max(1, (self.max_iter * n_samples) // self._batch_size)
```
To protect against `n_steps = 0`?

Reply:
at this point n_steps is guaranteed to be > 0: max_iter is checked to be > 0 and self._batch_size is defined to be less than n_samples.